pkg/calculator/calculator.go

```diff
@@ -21,26 +24,66 @@ type calculate struct {
 	totalEmbodied float64
 }
 
-// OperationalCPUEmissions are the emissions released from the machines the service is
+// TODO: This was borrowed from CCF calculations, do more research
+const memCoefficient = 0.000392 // Kilowatt Hour / Gigabyte Hour.
+
+// operationalEmissions determines the correct function to run to calculate the
+// operational emissions for the metric type
+func (c *calculate) operationalEmissions(metric *v1.Metric, interval time.Duration) (float64, error) {
+	switch metric.Name() {
+	case v1.CPU.String():
+		return c.cpu(metric, interval)
+	case v1.Memory.String():
+		return c.memory(metric, interval), nil
+	case v1.Storage.String():
+		return float64(-1), errors.New("error storage is not yet being calculated")
+	case v1.Network.String():
+		return float64(-1), errors.New("error networking is not yet being calculated")
+	default:
+		return 0, fmt.Errorf("error metric not found to be calculated: %+v", metric)
+	}
+}
+
+// cpu are the emissions released from the machines the service is
 // running on based on architecture and utilization.
-func (c *calculate) operationalCPUEmissions(interval time.Duration) float64 {
+func (c *calculate) cpu(m *v1.Metric, interval time.Duration) (float64, error) {
+	// Check that number of cores is set
+	if m.UnitAmount() == 0 {
+		return 0, errors.New("error Cores set to 0, this should never be the case")
+	}
+
 	// vCPUHours is the amount of cores on the machine multiplied by the interval of time
 	// for 1 hour. For example, if the machine has 4 cores and the interval of time is
 	// 5 minutes: The hourly time is 5/60 (0.083333333) * 4 cores = 0.333333333.
 	//nolint:unconvert //conversion to minutes does affect calculation
-	vCPUHours := c.cores * (float64(interval.Minutes()) / float64(60))
+	vCPUHours := m.UnitAmount() * (float64(interval.Minutes()) / float64(60))
 
 	// Average Watts is the average energy consumption of the service. It is based on
 	// CPU utilization and Minimum and Maximum wattage of the server. If the machine
 	// architecture is unknown the Min and Max wattage is the average of all machines
 	// for that provider, and is supplied in the provider defaults. This is being
 	// handled in the types/factors package (the point of reading in coefficient data).
-	avgWatts := c.minWatts + c.usageCPU*(c.maxWatts-c.minWatts)
+	if m.Usage() == 0 {
```

Comment:
Handleling validation in the calculation is actually a super smart move :thinking: fixes getters and setters problem

Reply:
The downside is that it does shift catching the errors further right in our development and at a later time. Slower feedback loop.